cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactional.java

```diff
@@ -0,0 +1,370 @@
+   *                                     or transaction exception from Tephra).
+   */
+  @Override
+  public void execute(TxRunnable runnable) throws TransactionFailureException {
+    executeLong(runnable, false);
+  }
+
+  @Nullable
+  TransactionInfo getTransactionInfo(String key) {
+    return transactionInfos.get(key);
+  }
+
+  void executeWithActiveOrLongTx(TxRunnable runnable) throws TransactionFailureException {
+    executeWithActiveOrLongTx(runnable, false);
+  }
+
+  void executeWithActiveOrLongTx(TxRunnable runnable, boolean asyncCommit) throws TransactionFailureException {
+    if (!tryExecuteWithActiveTransaction(runnable)) {
+      executeLong(runnable, asyncCommit);
+    }
+  }
+
+  /**
+   * Attempts to execute the given {@link TxRunnable} with an active transaction. If there is no active transaction,
+   * the {@link TxRunnable#run(DatasetContext)} will not be called.
+   *
+   * @param runnable the {@link TxRunnable} to run
+   * @return {@code true} if there is an active transaction and the {@link TxRunnable} was executed
+   * @throws TransactionFailureException if execute of the {@link TxRunnable} failed.
+   */
+  private boolean tryExecuteWithActiveTransaction(TxRunnable runnable) throws TransactionFailureException {
+    try {
+      TransactionalDatasetContext datasetContext = activeDatasetContext.get();
+      if (datasetContext == null || !datasetContext.canUse()) {
+        return false;
+      }
+
+      // Call the runnable
+      runnable.run(datasetContext);
+
+      // Persist all changes.
+      datasetContext.flush();
+
+      // Successfully execute with the current active transaction
+      return true;
+    } catch (Throwable t) {
+      Throwables.propagateIfInstanceOf(t, TransactionFailureException.class);
+      throw new TransactionFailureException("Failed to execute TxRunnable in transaction", t);
+    }
+  }
+
+  /**
+   * Executes the given runnable with a new long transaction. It will check for nested transaction and throws a
+   * {@link TransactionFailureException} if it finds one.
+   *
+   * @param runnable The {@link TxRunnable} to be executed inside a transaction
+   * @param asyncCommit if {@code true}, the transaction will be left open and it is expected to be committed
+   *                    by external entity asynchronously after this method returns;
+   *                    if {@code false}, the transaction will be committed when the runnable returned.
+   *
+   */
+  private void executeLong(TxRunnable runnable, boolean asyncCommit) throws TransactionFailureException {
+    TransactionalDatasetContext transactionalDatasetContext = activeDatasetContext.get();
+    if (transactionalDatasetContext != null) {
+      try {
+        if (!transactionalDatasetContext.commitOnJobEnded()) {
+          throw new TransactionFailureException("Nested transaction not supported. Active transaction is "
```

Comment:
Basically `commitOnJobEnded()` will return true for implicit transaction started on the client side and the transaction will be committed automatically when the job finished. This versus explicit transaction (by `Transactional.execute(TxRunnable)`), the transaction won't get committed when job ends (since there can be multiple jobs within the `TxRunnable.run()` method. If the current thread has a `!commitOnJobEnded` transaction, this means the user is trying to doing something like this:

```java
sec.execute(new TxRunnable() {
  public void run(DatasetContext context) {
    sec.execute(new TxRunnable() {      // <-- exception will be throw from here.
    });
  }
});